Userland/Libraries/LibUnicode/Normalize.cpp

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright (c) 2022, mat <???>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/Find.h>
+#include <AK/QuickSort.h>
+#include <LibUnicode/CharacterTypes.h>
+#include <LibUnicode/Normalize.h>
+#include <LibUnicode/UnicodeData.h>
+
+namespace Unicode {
+
+ALWAYS_INLINE static bool is_starter(u32 code_point)
+{
+    return Unicode::canonical_combining_class(code_point) == 0;
+}
+
+static u32 combine_code_points(u32 a, u32 b)
+{
+    Array<u32, 2> const points { a, b };
+    // FIXME: Find something better than linear search to find reverse mappings.
+    for (auto const& mapping : Unicode::code_point_decompositions()) {
+        if (mapping.tag == CompatibilityFormattingTag::Canonical && mapping.decomposition == points) {
+            if (code_point_has_property(mapping.code_point, Property::Full_Composition_Exclusion))
+                continue;
+            return mapping.code_point;
+        }
+    }
+    return 0;
+}
+
+enum class UseCompatibility {
+    Yes,
+    No
+};
+
+static void decompose_code_point(u32 code_point, Vector<u32>& output, UseCompatibility use_compatibility)
+{
+    // FIXME: Hangul characters are not handled.
+    auto const mapping = Unicode::code_point_decomposition(code_point);
+    if (mapping.has_value()) {
+        if (mapping->tag == CompatibilityFormattingTag::Canonical) {
+            decompose_code_point(mapping->decomposition[0], output, use_compatibility);
+            // "When a canonical mapping consists of a pair of characters,
+            //  the first character may itself be a character with a decomposition mapping,
+            //  but the second character never has a decomposition mapping."
+            if (mapping->decomposition.size() > 1 && mapping->decomposition[1] != 0) {
+                output.append(mapping->decomposition[1]);
+            }
+            return;
+        }
+        if (use_compatibility == UseCompatibility::Yes) {
+            for (auto code_point : mapping->decomposition) {
+                decompose_code_point(code_point, output, use_compatibility);
+            }
+            return;
+        }
+    }
+    output.append(code_point);
+}
+
+// This can be any sorting algorithm that maintains order (like std::stable_sort),
+// however bubble sort is the easier to implement, so go with it.
+template<typename T, typename LessThan>
+void bubble_sort(Span<T> span, LessThan less_than)
```

Comment:
Hmm we should make the single pivot quick sort stable if we can (not a comment for you, just talking out loud)